src/transform-sdk/rust/src/record.rs

```diff
@@ -0,0 +1,368 @@
+    key: Option<&'a [u8]>,
+    value: Option<&'a [u8]>,
+    timestamp: SystemTime,
+    headers: Vec<BorrowedHeader<'a>>,
+}
+
+impl<'a> BorrowedRecord<'a> {
+    /// Create a new record without any copies.
+    ///
+    /// NOTE: This method is useful for tests to mock out custom events to your transform function.
+    pub fn new(key: Option<&'a [u8]>, value: Option<&'a [u8]>, timestamp: SystemTime) -> Self {
+        Self {
+            key,
+            value,
+            timestamp,
+            headers: Vec::new(),
+        }
+    }
+
+    /// Create a new record without any copies.
+    ///
+    /// NOTE: This method is useful for tests to mock out custom events to your transform function.
+    pub fn new_with_headers(
+        key: Option<&'a [u8]>,
+        value: Option<&'a [u8]>,
+        timestamp: SystemTime,
+        headers: Vec<BorrowedHeader<'a>>,
+    ) -> Self {
+        Self {
+            key,
+            value,
+            timestamp,
+            headers,
+        }
+    }
+
+    pub(crate) fn read_from_payload(payload: &'a [u8], timestamp: SystemTime) -> Result<Self> {
+        let kv = read_kv(payload)?;
+        let mut headers: Vec<BorrowedHeader<'a>> = Vec::new();
+        let mut payload = &payload[kv.read..];
+        let header_count_result = varint::read(payload)?;
+        payload = &payload[header_count_result.read..];
+        for _ in 0..header_count_result.value {
+            let header_result = BorrowedHeader::read_from_payload(payload)?;
+            payload = &payload[header_result.read..];
+            headers.push(header_result.value);
+        }
+        let (key, value) = kv.value;
+        Ok(BorrowedRecord {
+            key,
+            value,
+            timestamp,
+            headers,
+        })
+    }
+
+    /// Returns the record's key or `None` if there is no key.
+    pub fn key(&self) -> Option<&'a [u8]> {
+        self.key
+    }
+
+    /// Returns the record's value or `None` if there is no value.
+    pub fn value(&self) -> Option<&'a [u8]> {
+        self.value
+    }
+
+    /// Returns the record's timestamp.
+    ///
+    /// NOTE: Record timestamps in Redpanda have millisecond resolution.
+    pub fn timestamp(&self) -> SystemTime {
+        self.timestamp
+    }
+
+    /// Return the headers for this record.
+    pub fn headers(&self) -> &[BorrowedHeader<'a>] {
+        &self.headers
+    }
+}
+
+/// Records may have a collection of headers attached to them.
+///
+/// Headers are opaque to the broker and are purely a mechanism for the producer and consumers to
+/// pass information.
+pub struct RecordHeader {
+    key: Option<Vec<u8>>,
+    value: Option<Vec<u8>>,
+}
+
+impl RecordHeader {
+    /// Create a new `RecordHeader`.
+    pub fn new(key: Option<Vec<u8>>, value: Option<Vec<u8>>) -> Self {
+        Self { key, value }
+    }
+
+    /// Create a new `RecordHeader` from a string pair.
+    pub fn new_from_string(key: &str, value: &str) -> Self {
```

Comment:
I don't see where this is used but presumingly it is for clients. Since you want owned byte sequences why not allow the client to give them to you? You are forcing an extra copy right now, but could instead just [`.into_bytes()`](https://doc.rust-lang.org/std/string/struct.String.html#method.into_bytes) on the moved string. 

It is cool to give users easy options to use more efficient APIs.

Example: `new_from_string(a, format!("bla {}", 42))`

Thinking about something like this

```rust
fn new_from_string<K: Into<Vec<u8>>, V: Into<Vec<u8>>>(k: K, v: V) -> Person {
       ...
}
```

Reply:
Way better, thank you really appreciate the review!